editoast/src/views/timetable/stdcm.rs

```diff
@@ -410,6 +425,60 @@ fn get_earliest_departure_time(
     )
 }
 
+// Return the latest time at which the train may arrive, margins included.
+fn get_latest_arrival_time(
+    data: &STDCMRequestPayload,
+    maximum_run_time_without_tolerance: u64,
+) -> DateTime<Utc> {
+    // Return the maximum time between:
+    //   * latest step time + its tolerance after
+    //   * start time + maximum start delay + max runtime
+    //   * first step time + its tolerance after + max runtime
+    let result_from_last_step: Option<DateTime<Utc>> = if let Some(PathfindingItem {
+        timing_data: Some(timing),
+        ..
+    }) = get_last_step_with_timing(data)
+    {
+        Some(
+            timing.arrival_time
+                + Duration::milliseconds(timing.arrival_time_tolerance_after as i64),
+        )
+    } else {
+        None
+    };
```

Comment:
Since you have `Option` all along the transformations, it might be better to express that series of operations in a more functional way (not tested so I might have miss something)?

```suggestion
    let result_from_last_step: Option<DateTime<Utc>> = get_last_step_with_timing(data)
      .and_then(|pathfinding_item| pathfinding_item.timing_data)
      .map(|timing| timing.arrival_time + Duration::milliseconds(timing.arrival_time_tolerance_after as i64));
```

Same below for `result_from_first_step`.

Reply:
My bad, the if let + pattern matching version is hard on the eyes indeed. Switched to mappings.